### PR TITLE
kprobe: fall back to tracefs if kernel doesn't support dots in PMU symbol

### PR DESCRIPTION
**Pull request by chenhengqi**
Centos 8 (kernel 4.18) has kprobe PMU, but don't allow `.` in symbol name,
add a new option to workaround this.

### CODE REVIEW

File: link/kprobe.go [hunk 1]
```diff
@@ -49,6 +49,13 @@ type KprobeOptions struct {
 	//
 	// Needs kernel 5.15+.
 	Cookie uint64
+
+	// Fallback to legacy tracefs kprobe interface
+	// event if kernel has kprobe PMU
+	//
+	// This allows old kernels which has kprobe PMU
+	// but don't support `.` in symbol names to work properly
```
Comment (mmat11) [inline link/kprobe.go:57]: are there kernels that supports PMU and don't support `.` in the symbol name and kernels that support PMU and `.` in the symbol name? (I am pretty sure on my kernel (5.16.x) I can attach uprobes PMU with `.` in the symbol)

If yes, what about implementing some feature detection to do this transparently? (kernel supports PMU && kernel doesn't support `.` -> go for tracefs)

File: link/kprobe.go [hunk 2]
```diff
@@ -173,7 +180,7 @@ func kprobe(symbol string, prog *ebpf.Program, opts *KprobeOptions, ret bool) (*
 	if err == nil {
 		return tp, nil
 	}
-	if err != nil && !errors.Is(err, ErrNotSupported) {
+	if err != nil && !errors.Is(err, ErrNotSupported) && !opts.Legacy {
```
Comment (mmat11) [inline link/kprobe.go:183]: opts can be `nil` here

File: link/kprobe.go (hunk 1 above)
Comment (chenhengqi) [reply link/kprobe.go:57]: > are there kernels that supports PMU and don't support . in the symbol name and kernels that support PMU and . in the symbol name?

Centos 8 with kernel v4.18. Did not find the related commit.

Comment (chenhengqi) [reply link/kprobe.go:57]: > If yes, what about implementing some feature detection to do this transparently? (kernel supports PMU && kernel doesn't support . -> go for tracefs)

Don't know how to reliably implement such detection. :(

**Comment by lmb** [comment]
@chenhengqi how does the centos kernel fail when `.` is in the name? In the best case we could add an `if errors.Is(err, EFOO)` case that enables the already existing tracefs fallback.

**Comment by chenhengqi** [comment]
> @chenhengqi how does the centos kernel fail when `.` is in the name? In the best case we could add an `if errors.Is(err, EFOO)` case that enables the already existing tracefs fallback.

The version of kernel shipped in CentOS 8 Linux is 4.18.

In 4.18, the kernel don't allow dot in symbol name in kprobe PMU case:

https://github.com/torvalds/linux/blob/94710cac0ef4ee177a63b5227664b38c95bbf703/kernel/trace/trace_kprobe.c#L340-L343

Can we always fallback to tracefs interface ? Or check the error against EINVAL ?

**Comment by ti-mo** [comment]
No need for feature detection IMO, falling back under a specific error like Lorenz suggested sounds like the best option to me. Is this possible?

**Comment by chenhengqi** [comment]
OK, just updated the code as you suggest.

File: link/kprobe.go [hunk 3]
```diff
@@ -335,7 +343,7 @@ func tracefsProbe(typ probeType, args probeArgs) (*perfEvent, error) {
 	}
 
 	// Get the newly-created trace event's id.
-	tid, err := getTraceEventID(group, args.symbol)
+	tid, err := getTraceEventID(group, sanitizedSymbol(args.symbol))
```
Comment (ti-mo) [inline link/kprobe.go:346]: Is this still necessary with the fallback to tracefs? I'm a bit foggy on this, but doesn't this modify the symbol name in a way that it no longer matches the kernel's? Couldn't we retry a PMU probe with a sanitized symbol in that case?

Comment (chenhengqi) [reply link/kprobe.go:346]: Yes, on line L396, we have to sanitize the `EVENT` part of kprobe_event.

Comment (ti-mo) [reply link/kprobe.go:346]: Okay, so this means the 'name' part of the event. That makes sense.

**Comment by ti-mo** [comment]
I've added a missing call to `sanitizedSymbol()` in `closeTraceFSProbeEvent()`. Closing a tracefs event created with a sanitized name would of course fail otherwise.

I've also converted `TestK(ret)probe` to be table-driven tests and found some kernel symbols with local visibility that were present in kernel versions ranging from 4.4 to 5.15, so this fallback behaviour can actually be exercised.

**Comment by chenhengqi** [comment]
Thanks.

**Pull request merged**